### PR TITLE
Split CI providers test

### PR DESCRIPTION
**Pull request by troopa81**
## Description

Spersedes #41473

It all started with my will to restore the Oracle provider tests in CI #41473

But the Oracle docker is too big (around 13GB), and with QGIS build and dependencies exceeds the GitHub runner size [limitation](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources). 

To fix this, I choose to run the providers tests in parallel in 4 different runners (thanks @3nids for the artifact hint): 
- Oracle tests
- HANA tests
- PostGres tests
- All other non provider tests

I use docker multi build stage in qgis-build-deps.dockerfile to build a docker with and without dependencies to lower the size of the tests running docker.

Funded by Lille Metropole
cc @Jean-Roc

### CODE REVIEW

File: tests/src/python/CMakeLists.txt
Comment (nyalldawson) [inline tests/src/python/CMakeLists.txt:368]: These two shouldn't be dependent on the postgres tests being enabled. You'll need to split the tests themselves into two seperate files so all the non postgres parts of the files can still be run.

Comment (troopa81) [reply tests/src/python/CMakeLists.txt]: Done for PyQgsVectorFileWriter. On the other hand, all tests in PyQgsQueryResultModel needs PostGres

Comment (nyalldawson) [reply tests/src/python/CMakeLists.txt]: Ah sorry, I was referring to PyQgsVectorLayerUtils

File: .github/workflows/run-tests.yml
Comment (3nids) [inline .github/workflows/run-tests.yml:125]: can we integrate the ctest line in the bash file?

File: .github/workflows/run-tests.yml
Comment (3nids) [inline .github/workflows/run-tests.yml:129]: could you maybe introduce a variable to define if we run the tests or no, ideally in the strategy?

File: .github/workflows/run-tests.yml
Comment (3nids) [inline .github/workflows/run-tests.yml:264]: could you add a remark here that is duplicate code from above and that both need to be updated (put the remark at the 2 places)

**Review by 3nids — COMMENTED** [review]
excellent! I really like where this is going.
Small remarks in the mean time.

**Comment by 3nids** [comment]
If I understand well, we will get several dash reports each time, am I right?

**Comment by troopa81** [comment]
> If I understand well, we will get several dash reports each time, am I right?

Yes, when it fails you bet dash report with _ORACLE, _POSTGRES, _ALL_BUT_PROVIDERS

**Comment by 3nids** [comment]
> Yes, when it fails you bet dash report with _ORACLE, _POSTGRES, _ALL_BUT_PROVIDERS

The reports are pushed in any case. But I don't see this an issue.
@jef-n does it sound ok to you too?

**Comment by m-kuhn** [comment]
CC @stefanuhrig @mrylov

File: .github/workflows/run-tests.yml
Comment (troopa81) [reply .github/workflows/run-tests.yml:264]: Done

File: .github/workflows/run-tests.yml
Comment (troopa81) [reply .github/workflows/run-tests.yml]: Done 72e66ee

File: .github/workflows/run-tests.yml
Comment (troopa81) [reply .github/workflows/run-tests.yml]: Done

File: tests/src/python/CMakeLists.txt
Comment (troopa81) [reply tests/src/python/CMakeLists.txt]: thanks, It's done too

File: .docker/docker-qgis-build.sh
Comment (3nids) [inline .docker/docker-qgis-build.sh:87]: do you think we still the need the definition of `CTEST_BUILD_COMMAND` in `.docker/docker-variables.env` and in `.ci/config.ctest`?

**Comment by 3nids** [comment]
thanks for the update!

File: tests/src/python/CMakeLists.txt
Comment (elpaso) [inline tests/src/python/CMakeLists.txt:392]: Can you move the tests that use PG from PyQgsServerWMSGetFeatureInfoto PyQgsServerWMSGetFeatureInfoPG ?

Comment (troopa81) [reply tests/src/python/CMakeLists.txt]: Done

**Comment by jef-n** [comment]
> The reports are pushed in any case. But I don't see this an issue.
> @jef-n does it sound ok to you too?

I don't have a problem with more results - I would esp. like to get back the nightly results.  But I didn't find time to explore why they aren't uploaded anymore (or why there are occasionally exceptions).

But orfeo maybe?

File: .docker/docker-qgis-build.sh
Comment (troopa81) [reply .docker/docker-qgis-build.sh:87]: I don't know, I removed it. Will see if it fails without it

**Comment by troopa81** [comment]
I have to add another stage with binary only needed for oracle test in qgis-build-deps.dockerfile to give enough free space to Oracle docker.

I add some logs to monitor disk space if an error occured during running test. 

I'm checking few details, don't merge for now.

File: .docker/docker-qgis-build.sh
Comment (troopa81) [reply .docker/docker-qgis-build.sh:87]: It is definitely needed. I restored it

**Comment by troopa81** [comment]
Everything is green but I think required build (20.04) and build (20.10) are not matching the new build name because there is the strategy boolean in it. 

Plus the new "run tests" should be set as required, I think it needs to be done after merging in branch protection rule.

**Comment by 3nids** [comment]
So ready for merge?

**Comment by troopa81** [comment]
@3nids to me yes :)

**Review by 3nids — APPROVED** [review]

**Comment by 3nids** [comment]
I have update the branch protection for master
not sure if exsisting PRs need to be closed - reopened.

**Comment by nyalldawson** [comment]
@troopa81 the oracle test is failing semi-regularly at the moment, eg https://github.com/qgis/QGIS/pull/42697/checks?check_run_id=2289254690 . Do you know why this is?

**Comment by troopa81** [comment]
> @troopa81 the oracle test is failing semi-regularly at the moment, eg https://github.com/qgis/QGIS/pull/42697/checks?check_run_id=2289254690 . Do you know why this is?

Nope. I look into it

**Comment by troopa81** [comment]
I fail to reproduce in local. It seems to crash in oci library... 

I'm currently trying to run tests in my repo with logs several times, see if I can make it crash.

**Pull request merged**
